File: lib/app/models/application_entity/application_entity.ex
defmodule ApplicationEntity do
  @moduledoc """
    Application entity is any generic enity capable of being an actor in financial transaction
  """

  @typedoc """
    `application_entity.pub_id` db reference
  """
  @type id :: String.t()

  @typedoc """
    `application_entity.external_id field
  """
  @type external_id :: String.t()

  @typedoc """
    Type of application entity
  """
  @type type :: :CLIENT | :MASTER

  @type t :: %ApplicationEntity{
          id: id(),
          type: type(),
          external_id: external_id()
        }

  defstruct id: nil,
            type: nil,
            external_id: nil

  @spec find_by_external_id(ApplicationEntity.external_id()) :: ApplicationEntity.id()
  def find_by_external_id(id) do
    id
    |> DB.query_val("SELECT pub_id FROM application_entity WHERE external_id = $1")
  end
end

File: pkg/models/application_entity/application_entity.go
package models

// `application_entity.pub_id` db reference
type ApplicationEntityId string

// `application_entity.pub_id` db reference
type ApplicationEntityExternalId string

// Type of application entity
type ApplicationEntityType string

const (
	Client ApplicationEntityType = "CLIENT"
	Master ApplicationEntityType = "MASTER"
)

// Application entity is any generic enity capable of being an
// actor in financial transaction
type ApplicationEntity struct {
	Id         ApplicationEntityId
	Type       ApplicationEntityType
	ExternalId ApplicationEntityExternalId
}

func FindByExternalId(id ApplicationEntityExternalId) ApplicationEntityId {
	db.QueryVal("SELECT pub_id FROM application_entity WHERE external_id = $1", id)
}
